### ISSUE TITLE
Enforce canonical CBOR encodings

### ISSUE THREAD

**Issue by mrBliss** [input-output-hk/ouroboros-network#905]
Quoting @dcoutts:
> Our new impl in general does not check for canonical encodings because it doesn't need to care any more (yay).
But the old impl does care. So to avoid an unfortunate disagreement with the old nodes we should still make sure that new stuff uses canonical encodings.
We still enforce canonical encodings when decoding addresses, but not elsewhere.
I think the only other place we really have to enforce this for the Byron rules is in accepting new transactions. We don't need to worry about blocks or other things like update proposals, votes etc because those are produced by the federated nodes under our control. So it's only txs that come in from users where we need to check we don't accept txs with non-canonical encodings and put them into blocks.
So my thought is that we do a canonical check when we're adding new txs to the mempool and nowhere else. And we do that canonical check generically, rather than by changing our cbor decoders.
The canonical check is simple: re-encode and check the bytes are the same. This is more expensive, but if we're only doing it in the mempool for new txs then that's fine.
And where should it go? Is the right place in the ledger function that is used by the mempool, or should it go where the function is used by the mempool?

**Comment by mrBliss** [comment]
To check whether a transaction is canonical, we need access to the original bytes so we can compare them with a fresh, canonical encoding of the transaction.

Let's look at the current situation. We are using the following type for generalised transactions:
```haskell
data GenTx (ByronBlockOrEBB cfg) = ByronTx
  { byronTx   :: !(CC.UTxO.ATxAux ByteString)
  , byronTxId :: CC.UTxO.TxId
    -- ^ This field is lazy on purpose so that the TxId is computed on
    -- demand.
  }
```
We decode it using:
```haskell
decodeByronGenTx :: Decoder s (GenTx (ByronBlockOrEBB cfg))
decodeByronGenTx =
    mkByronTx . annotate <$> fromCBOR
  where
    annotate :: CC.UTxO.ATxAux ByteSpan -> CC.UTxO.ATxAux ByteString
    annotate CC.UTxO.ATxAux{aTaTx, aTaWitness} =
      CC.UTxO.ATxAux{
        aTaTx      = reAnnotate aTaTx,
        aTaWitness = reAnnotate aTaWitness
      }
```
Note the use of `reAnnotate`, which throws away the annotation (the bytes) and reserialises the given value to use as its new (and thus canonical) annotation. We will compute the hash of the `ATxAux` in `mkByronTx` using the hash of the `aTaTx` field (the witness is ignored?), which actually involves serialising *again*. So we're serialising twice *and* losing access to the original bytes! The upside is that we always have a canonical encoding :grin:.  I'm not 100% sure, but I suspect there's a danger that the sender of the transaction has a different hash (`TxId`) for the transaction than we have, which is not what we want.

I think the next step to take is to avoid throwing away the original bytes, which will make it possible to do the canonical check, and which will also improve performance as we will avoid two serialisations per incoming transaction.

I have been told the ledger team are planning to rework the annotations. I hope that will improve things.

P.S. #999 will not fundamentally change the situation, it will just add more constructors with types that have the same problem.

**Comment by mrBliss** [comment]
Conclusion of a discussion with @edsko and @dnadales: *the ledger will [detect and reject](https://github.com/input-output-hk/cardano-ledger/blob/f8e802b24175b7a20f13f066f9cce32716c9d414/cardano-ledger/src/Cardano/Chain/UTxO/Validation.hs#L274) transactions with non-canonical encodings*. It will compute the hash of the `ByteString` annotation of the transaction and compare that with the signed hash that is stored in the transaction witness. If the hashes do not match, it will return a `TxValidationWitnessWrongSignature` error. Although the error does not explicitly say the transaction is invalid because of a non-canonical encoding, this is good enough.

It is **important** to make sure the *`ByteString` annotation of the transaction is canonical*. If the non-canonical encoding we received via the network is used as the annotation, then the hash will match and the transaction will be accepted. So the solution is to serialise each incoming transaction from scratch and to reuse that encoding, which must be canonical, as the annotation.

**Comment by dnadales** [comment]
> Conclusion of a discussion with @edsko and @dnadales: the ledger will detect and reject transactions with non-canonical encodings.

@mrBliss I think we should be careful here. As you pointed out later on in the paragraph, the ledger cannot detect (without de-serializing with some sort of "canonical decoder") non-canonical encodings.

**Comment by mrBliss** [comment]
Addressed by #999.
